File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();

        // flip matrix by its diagonal
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                swap(matrix[i][j], matrix[j][i]);
            }   
        }

        // flip matrix by straight line
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < int(n / 2); ++j) {
                swap(matrix[i][j], matrix[i][n - j - 1]);
            }
        }
        return;
    }
};

File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # swap along x asix and y asix
        n = len(matrix)
        for i in range(n):
            for j in range(i + 1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

        for row in matrix:
            left = 0
            right = n - 1

            while left < right:
                row[left], row[right] = row[right], row[left]
                left += 1
                right -= 1



        